chromium/browser_type.go

```diff
@@ -173,37 +195,52 @@ func (b *BrowserType) Name() string {
 	return "chromium"
 }
 
+func (b *BrowserType) createTempDir(
+	dirPath interface{},
+	logger *log.Logger,
+) (finalDirPath string, c common.CleanupFunc, rerr error) {
+	// use the provided directory or create a temporary one.
+	if err := b.storage.Make("", dirPath); err != nil {
+		return "", nil, fmt.Errorf("cannot make temp data directory: %w", err)
+	}
+
+	finalDirPath = b.storage.Dir
+
+	m := sync.Mutex{}
+	return finalDirPath, func() {
+		m.Lock()
+		defer m.Unlock()
+		if err := b.storage.Cleanup(); err != nil {
```

Comment:
What is this mutex protecting against? Concurrent calls to `Cleanup()`? What would actually fail if that were to happen?

I tried removing it and calling this lambda concurrently, and couldn't reproduce a failure. But if it can fail, then maybe it should be an implementation detail in `Cleanup()` instead?

Reply:
It is protecting an implementation detail within `Cleanup()`, so i will move it into `Cleanup()`.

Odd that you weren't able to find the race condition when running it concurrently. I ran all the tests with the `--race` flag to help me spot it. It's because of:

```
	if d.fsRemoveAll == nil {
		d.fsRemoveAll = os.RemoveAll
	}

        return d.fsRemoveAll(d.Dir)
```

I could just make it:

```
	return os.RemoveAll(d.Dir)
```

I don't know the context behind why we're doing the way we're doing it.

Reply:
Ah, I see.

> I don't know the context behind why we're doing the way we're doing it.

I'm not sure about it myself. In general, it's a good idea to abstract away FS function calls so that you can mock them out in tests and avoid touching the FS (or use a virtual FS like afero), but from what I've seen we're not using them in our tests, or anywhere else.

@inancgumus originally wrote this, so WDYT? Should we remove `fsMkdirTemp` and `fsRemoveAll` and just use `os` directly?

Reply:
Can we use `sync.Once` here since `fsRemoveAll` and friends are initialized only in the constructor.

Reply:
Oh, good point on virtual filesystems. I will give that a go as that will make things easier to deal with if the tests fails. Currently, if the directory removal test fails, it will cause future runs of the tests to fail as it only deletes the directory that's created in the current run.

I will use `sync.Once` to init the functions.

Reply:
I will work on this refactor in issue grafana/xk6-browser#404